opensquirrel/utils/singleton.py

```diff
@@ -0,0 +1,14 @@
+from __future__ import annotations
+
+import typing as t
+
+_T = t.TypeVar("_T")
+
+
+class Singleton(type, t.Generic[_T]):
+    _instances: t.ClassVar[dict[type, t.Any]] = {}
+
+    def __call__(cls, *args: t.Any, **kwargs: t.Any) -> _T:
+        if cls not in cls._instances:
+            cls._instances[cls] = super().__call__(*args, **kwargs)
+        return t.cast(_T, cls._instances[cls])
```

Comment:
Wow, what is going on here? Looks impressive 😅 Can you explain a bit what is going on?

Is this something that can imported?

Reply:
To be honest:

- I just copy-pasted a solution from StackOverflow. There are many implementations of a Singleton in Python, but it seems the one using metaclasses is the one with more pros and less cons.
- From there, again using StackOverflow, I found a new version with type annotations that would pass `mypy --strict`.
- Finally, I used ChatGPT to make that code pass `ruff check --fix`
important thing is that:
- We now have a Singleton class in `utils/singleton.py` that we can reuse for other occassions.

And yes, definitely, we can reuse this Singleton class from `utils/singleton.py` if we need more classes in the codebase to follow the Singleton pattern.

In general, I would read that code as:
- I'm a Singleton class, and you can inherit from me if you want to implement a class following the Singleton pattern.
- What does that mean? That there is going to be only one instance of that subclass during runtime.
- How will I do it?
  - First, I will keep a dictionary where the key is the subclass type (e.g., InstructionLibrary) and the value is an instance of that subclass type.
  - Then, I will implement the __call__ method, so you will be able to use your subclass as a functor, e.g., InstructionLibrary(), to get the single instance of your subclass.
  - Finally, how will I implement that __call__ method? If I have previously created an instance of your subclass, I will just return it, otherwise I will create it, save it in the internal dictionary, and return it.

Reply:
Thanks for the detailed explanation!

I would propose to import the classes from typing (consistent with how we do this in the rest of the code):

```suggestion
from typing import TypeVar, Generic, ClassVar, Any, cast

_T = TypeVar("_T")

class Singleton(type, Generic[_T]):
    _instances: ClassVar[dict[type, Any]] = {}

    def __call__(cls, *args: Any, **kwargs: Any) -> _T:
        if cls not in cls._instances:
            cls._instances[cls] = super().__call__(*args, **kwargs)
        return cast(_T, cls._instances[cls])
```

Reply:
Done!